catalyst_voices_packages/catalyst_cardano_serialization/lib/src/rbac/role_based_access_control.dart

```diff
@@ -0,0 +1,354 @@
+import 'package:catalyst_cardano_serialization/catalyst_cardano_serialization.dart';
+import 'package:cbor/cbor.dart';
+import 'package:equatable/equatable.dart';
+
+/// Defines the X509 Role Based Access Control transaction metadata.
+final class RbacMetadata extends Equatable {
+  /// Un-ordered List of DER encoded x509 certificates.
+  final List<X509DerCertificate>? derCerts;
+
+  /// Un-ordered List of CBOR encoded C509 certificates
+  /// (or metadatum references).
+  //
+  // TODO(dtscalac): support C509CertInMetadatumReference
+  final List<C509Certificate>? cborCerts;
+
+  /// Ordered list of simple public keys that are registered.
+  final List<Ed25519PublicKey>? publicKeys;
+
+  /// Revocation list of certs being revoked by an issuer.
+  final List<CertificateHash>? revocationSet;
+
+  /// Set of role registration data.
+  final Set<RoleData>? roleDataSet;
+
+  /// The default constructor for [RbacMetadata].
+  const RbacMetadata({
+    this.derCerts,
+    this.cborCerts,
+    this.publicKeys,
+    this.revocationSet,
+    this.roleDataSet,
+  });
+
+  /// Deserializes the type from cbor.
+  factory RbacMetadata.fromCbor(CborValue value) {
+    final map = value as CborMap;
+    final derCerts = map[const CborSmallInt(10)] as CborList?;
+    final cborCerts = map[const CborSmallInt(20)] as CborList?;
+    final publicKeys = map[const CborSmallInt(30)] as CborList?;
+    final revocationSet = map[const CborSmallInt(40)] as CborList?;
+    final roleDataSet = map[const CborSmallInt(50)] as CborList?;
+
+    return RbacMetadata(
+      derCerts: derCerts?.map(X509DerCertificate.fromCbor).toList(),
+      cborCerts: cborCerts?.map(C509Certificate.fromCbor).toList(),
+      publicKeys: publicKeys?.map(Ed25519PublicKey.fromCbor).toList(),
+      revocationSet: revocationSet?.map(CertificateHash.fromCbor).toList(),
+      roleDataSet: roleDataSet?.map(RoleData.fromCbor).toSet(),
+    );
+  }
+
+  /// Serializes the type as cbor.
+  CborValue toCbor() {
+    return CborMap({
+      if (derCerts != null && derCerts!.isNotEmpty)
+        const CborSmallInt(10): CborList([
+          for (final cert in derCerts!) cert.toCbor(),
+        ]),
+      if (cborCerts != null && cborCerts!.isNotEmpty)
+        const CborSmallInt(20): CborList([
+          for (final cert in cborCerts!) cert.toCbor(),
+        ]),
+      if (publicKeys != null && publicKeys!.isNotEmpty)
+        const CborSmallInt(30): CborList([
+          for (final key in publicKeys!) key.toCbor(),
+        ]),
+      if (revocationSet != null && revocationSet!.isNotEmpty)
+        const CborSmallInt(40): CborList([
+          for (final revocation in revocationSet!) revocation.toCbor(),
+        ]),
+      if (roleDataSet != null && roleDataSet!.isNotEmpty)
+        const CborSmallInt(50): CborList([
+          for (final roleData in roleDataSet!) roleData.toCbor(),
+        ]),
+    });
+  }
```

Comment:
What do you think if we can replace the if statements with a more functional approach:

something like this:
```dart
  CborList? _createCborList<T>(List<T>? items, CborValue Function(T) toCbor) {
    if (items == null || items.isEmpty) return null;
    return CborList(items.map(toCbor).toList());
  }

  /// Builds a CborMap from the class properties.
  Map<CborSmallInt, CborList> _buildCborMap() {
    final entries = [
      MapEntry(
        const CborSmallInt(10),
        _createCborList<X509DerCertificate>(
          derCerts,
          (item) => item.toCbor(),
        ),
      ),
      MapEntry(
        const CborSmallInt(20),
        _createCborList<C509Certificate>(
          cborCerts,
          (item) => item.toCbor(),
        ),
      ),
      MapEntry(
        const CborSmallInt(30),
        _createCborList<Ed25519PublicKey>(
          publicKeys,
          (item) => item.toCbor(),
        ),
      ),
      MapEntry(
        const CborSmallInt(40),
        _createCborList<CertificateHash>(
          revocationSet,
          (item) => item.toCbor(),
        ),
      ),
      MapEntry(
        const CborSmallInt(50),
        _createCborList<RoleData>(
          roleDataSet?.toList(),
          (item) => item.toCbor(),
        ),
      ),
    ];

    return Map.fromEntries(
      entries
          .where((entry) => entry.value != null)
          .map((entry) => MapEntry(entry.key, entry.value!)),
    );
  }

  CborValue toCbor() => CborMap(_buildCborMap());
````

Reply:
Ok, slightly modified it to make it shorter:
```dart
  /// Serializes the type as cbor.
  CborValue toCbor() => CborMap(_buildCborMap());

  /// Builds a CborMap from the class properties.
  Map<CborSmallInt, CborValue> _buildCborMap() {
    final entries = <CborSmallInt, CborValue?>{
      const CborSmallInt(10): _createCborList<X509DerCertificate>(
        derCerts,
        (item) => item.toCbor(),
      ),
      const CborSmallInt(20): _createCborList<C509Certificate>(
        cborCerts,
        (item) => item.toCbor(),
      ),
      const CborSmallInt(30): _createCborList<Ed25519PublicKey>(
        publicKeys,
        (item) => item.toCbor(),
      ),
      const CborSmallInt(40): _createCborList<CertificateHash>(
        revocationSet,
        (item) => item.toCbor(),
      ),
      const CborSmallInt(50): _createCborList<RoleData>(
        roleDataSet?.toList(),
        (item) => item.toCbor(),
      ),
    }..removeWhere((key, value) => value == null);

    return entries.cast();
  }

  CborList? _createCborList<T>(List<T>? items, CborValue Function(T) toCbor) {
    if (items == null || items.isEmpty) return null;
    return CborList(items.map(toCbor).toList());
  }
```